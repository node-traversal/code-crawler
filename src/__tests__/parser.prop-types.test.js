const _ = require('lodash');
const fs = require('fs');
const path = require('path');
const parser = require('../parser');
const reactPlugin = require('../parser-plugin-react');

const basePath = process.cwd();
const examplesPath = path.join(basePath, 'examples/components');

describe('The react parser plugin should', () => {
    const options = {
        basePath: examplesPath,
        packageName: '@my-project/the-package'
    };

    const parse = (file) => {
        const filePath = path.join(examplesPath, file);
        const code = fs.readFileSync(filePath, 'utf-8');
        return parser.parserReactCode(code, _.assign({}, options, { filePath }), [ reactPlugin ]);
    };

    it('parse simple react code', () => {
        expect(parse('simple-example.js')).toMatchSnapshot();
    });

    it('parse pre-es6 react code', () => {
        expect(parse('without-es6.js')).toMatchSnapshot();
    });

    it('parse hook react code', () => {
        expect(parse('draftjs.js')).toMatchSnapshot();
    });

    it('parse scoped react code', () => {
        expect(parse('scoped-package.js')).toMatchSnapshot();
    });

    it('parse simple react prop-types', () => {
        expect(parse('prop-types-simple.js')).toMatchSnapshot();
    });

    it('parse react prop-types with single argument', () => {
        expect(parse('prop-types-single-arg.js')).toMatchSnapshot();
    });

    it('parse react prop-types with enum arguments', () => {
        expect(parse('prop-types-enum-args.js')).toMatchSnapshot();
    });

    it('parse react prop-types with shape', () => {
        expect(parse('prop-types-shape.js')).toMatchSnapshot();
    });

    it('parse react prop-types with required specifier', () => {
        expect(parse('prop-types-required.js')).toMatchSnapshot();
    });

    it('parse complex react prop-types', () => {
        expect(parse('complex-prop-types.js')).toMatchSnapshot();
    });
});
