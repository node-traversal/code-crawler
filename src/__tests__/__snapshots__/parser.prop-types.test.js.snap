// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The react parser plugin should parse complex react prop-types 1`] = `
Object {
  "deps": Array [
    Object {
      "name": "react",
      "var": "React",
    },
    Object {
      "name": "prop-types",
      "var": "PropTypes",
    },
  ],
  "props": Array [
    Object {
      "name": "optionalArray",
      "type": "array",
    },
    Object {
      "name": "optionalBool",
      "type": "bool",
    },
    Object {
      "name": "optionalFunc",
      "type": "func",
    },
    Object {
      "name": "optionalNumber",
      "type": "number",
    },
    Object {
      "name": "optionalObject",
      "type": "object",
    },
    Object {
      "name": "optionalString",
      "type": "string",
    },
    Object {
      "name": "optionalSymbol",
      "type": "symbol",
    },
    Object {
      "name": "optionalNode",
      "type": "node",
    },
    Object {
      "name": "optionalElement",
      "type": "element",
    },
    Object {
      "name": "optionalElementType",
      "type": "elementType",
    },
    Object {
      "args": "Message",
      "name": "optionalMessage",
      "type": "instanceOf",
    },
    Object {
      "args": Array [
        "News",
        "Photos",
      ],
      "name": "optionalEnum",
      "type": "oneOf",
    },
    Object {
      "args": Array [
        "string",
        "number",
        "{\\"type\\":\\"CallExpression\\",\\"callee\\":{\\"type\\":\\"MemberExpression\\",\\"start\\":1424,\\"end\\":1444,\\"loc\\":{\\"start\\":{\\"line\\":45,\\"column\\":8},\\"end\\":{\\"line\\":45,\\"column\\":28}},\\"object\\":{\\"type\\":\\"Identifier\\",\\"start\\":1424,\\"end\\":1433,\\"loc\\":{\\"start\\":{\\"line\\":45,\\"column\\":8},\\"end\\":{\\"line\\":45,\\"column\\":17},\\"identifierName\\":\\"PropTypes\\"},\\"name\\":\\"PropTypes\\"},\\"property\\":{\\"type\\":\\"Identifier\\",\\"start\\":1434,\\"end\\":1444,\\"loc\\":{\\"start\\":{\\"line\\":45,\\"column\\":18},\\"end\\":{\\"line\\":45,\\"column\\":28},\\"identifierName\\":\\"instanceOf\\"},\\"name\\":\\"instanceOf\\"},\\"computed\\":false},\\"arguments\\":[{\\"type\\":\\"Identifier\\",\\"start\\":1445,\\"end\\":1452,\\"loc\\":{\\"start\\":{\\"line\\":45,\\"column\\":29},\\"end\\":{\\"line\\":45,\\"column\\":36},\\"identifierName\\":\\"Message\\"},\\"name\\":\\"Message\\"}]}",
      ],
      "name": "optionalUnion",
      "type": "oneOfType",
    },
    Object {
      "args": "number",
      "name": "optionalArrayOf",
      "type": "arrayOf",
    },
    Object {
      "args": "number",
      "name": "optionalObjectOf",
      "type": "objectOf",
    },
    Object {
      "args": Array [
        Object {
          "name": "optionalProperty",
          "type": "string",
        },
        Object {
          "name": "requiredProperty",
          "required": true,
          "type": "number",
        },
      ],
      "name": "optionalObjectWithShape",
      "type": "shape",
    },
    Object {
      "args": Array [
        Object {
          "name": "optionalProperty",
          "type": "string",
        },
        Object {
          "name": "requiredProperty",
          "required": true,
          "type": "number",
        },
      ],
      "name": "optionalObjectWithStrictShape",
      "type": "exact",
    },
    Object {
      "name": "requiredFunc",
      "required": true,
      "type": "func",
    },
    Object {
      "name": "requiredAny",
      "required": true,
      "type": "any",
    },
  ],
}
`;

exports[`The react parser plugin should parse hook react code 1`] = `
Object {
  "deps": Array [
    Object {
      "name": "react",
      "var": "React",
    },
    Object {
      "name": "react-dom",
      "var": "ReactDOM",
    },
    Object {
      "name": "draft-js",
      "var": "Editor",
    },
    Object {
      "name": "draft-js",
      "var": "EditorState",
    },
  ],
}
`;

exports[`The react parser plugin should parse pre-es6 react code 1`] = `
Object {
  "deps": Array [
    Object {
      "name": "create-react-class",
      "var": "createReactClass",
    },
  ],
}
`;

exports[`The react parser plugin should parse react prop-types with enum arguments 1`] = `
Object {
  "props": Array [
    Object {
      "args": Array [
        "News",
        "Photos",
      ],
      "name": "optionalEnum",
      "type": "oneOf",
    },
  ],
}
`;

exports[`The react parser plugin should parse react prop-types with required specifier 1`] = `
Object {
  "props": Array [
    Object {
      "name": "requiredAny",
      "required": true,
      "type": "any",
    },
  ],
}
`;

exports[`The react parser plugin should parse react prop-types with shape 1`] = `
Object {
  "props": Array [
    Object {
      "args": Array [
        Object {
          "name": "firstName",
          "type": "string",
        },
        Object {
          "name": "age",
          "type": "number",
        },
      ],
      "name": "objectWithShape",
      "type": "shape",
    },
  ],
}
`;

exports[`The react parser plugin should parse react prop-types with single argument 1`] = `
Object {
  "props": Array [
    Object {
      "args": "Message",
      "name": "optionalMessage",
      "type": "instanceOf",
    },
  ],
}
`;

exports[`The react parser plugin should parse scoped react code 1`] = `
Object {
  "deps": Array [
    Object {
      "name": "react",
      "var": "React",
    },
    Object {
      "name": "@material-ui/core/Button",
      "var": "Button",
    },
  ],
}
`;

exports[`The react parser plugin should parse simple react code 1`] = `
Object {
  "deps": Array [
    Object {
      "name": "react",
      "var": "React",
    },
    Object {
      "name": "prop-types",
      "var": "PropTypes",
    },
  ],
  "props": Array [
    Object {
      "name": "firstName",
      "type": "string",
    },
  ],
}
`;

exports[`The react parser plugin should parse simple react prop-types 1`] = `
Object {
  "props": Array [
    Object {
      "name": "firstName",
      "type": "string",
    },
  ],
}
`;
